### PR TITLE
span: tighten up range based lookup spans

### PR DESCRIPTION
**Pull request by yuzefovich**
**rowexec: clean up locality optimized span generator**

Previously, the locality optimized span generator would always check
whether the local span generator has any matching rows for the given
key, and if that returned nothing, then it would check the remote span
generator. This works ok because for all remotely looked up rows the
prefix of the key would be different from all prefixes of the local
spans.

However, this can be improved if we take advantage of the way the join
reader uses the lookup spans and processes the looked rows. Namely, in
case of a locality optimized search it does the following:
- generate local spans
- lookup all local spans
- process all returned rows (here we will get the matching input row
indices from the span generator)
- if some input rows didn't get a match, then generate remote spans
- lookup all remote spans
- process all returned rows (getting the matching indices again).

This outline makes it clear that the span generator is aware of what
spans (local or remote) were returned last when it is asked for the
matching row indices, so this commit makes the corresponding change.

This makes the code cleaner and avoids an unnecessary map lookup for
remotely looked up rows.

Release note: None

**execbuilder: add some tests exposing inefficiencies in lookup join spans**

This commit adds some tests that examine the trace of the lookup joins
to expose that some keys are fetched when they don't pass the inequality
filter. In other words, the tests expose that the lookup join spans are
not as tight as possible which results in unnecessary fetches.

Release note: None

**span: tighten up range based lookup spans**

This commit tightens up the spans produced for range-based lookup joins
(where we have a single inequality) in some cases. Previously, we could
lookup more rows that needed, and they would get discarded later because
the partial key for not-needed looked up rows would not be contained by
the lookup spans.

This commit makes the following adjustments:
- if we have a start datum and we want it to be excluded, then we
"advance" the datum using `Next`/`Prev`.
- we now correctly skip the nulls.

The complication is that `Next`/`Prev` are invalid for some datums (for
example, for the "max"/"min" of their type or for some types - like
decimals - altogether). In order to go around this problem the optimizer
will no longer use range-based lookups when these functions cannot be
used on the span boundary datums.

I believe this limitation is acceptable given that it effectively
disables range-based lookups and falls back to the ON expression
evaluation in very limited set of circumstances, but it will allow us to
clean up and speed up the lookup join execution in the follow-up
commits.

Release note: None

### CODE REVIEW

**Review by yuzefovich — COMMENTED** [review]
This is still somewhat WIP, but I wanted to make sure I'm not missing something before starting looking for edge cases.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-:-N06cuLI3lWQBi5sI6Q4:b-r5oie9)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach and @rytaft)

___
*[pkg/sql/span/span_builder.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N06d-rY6oFd4CMb6HAl:-N06d-rY6oFd4CMb6HAm:b-ayfhaw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ba6bf4cc821550d6991cf900c4cd5528eae98aa/pkg/sql/span/span_builder.go#L108)):*
> ```Go
> 			// inclusive, yet our inequality condition wants to exclude it,
> 			// we'll "advance" the start datum by one.
> 			// TODO(yuzefovich): make sure that the optimizer is able to detect
> ```

Curious if Becca or Tommy have any insight on this.
___
*[pkg/sql/span/span_builder.go, line 114 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N06dAjsACpwzLgOiaed:-N06dAjsACpwzLgOiaee:b-53cxkk) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dafee332eef692a753ec3a5459294c7b90a10df/pkg/sql/span/span_builder.go#L114)):*
> ```Go
> 			startKey = encoding.EncodeNullAscending(startKey)
> 		}
> 		startKey = startKey.Next()
> ```

I couldn't understand why we do this - Tommy, do you remember?

**Comment by cucaroach** [comment]
> I couldn't understand why we do this - Tommy, do you remember?

Not really, I do remember that it was necessary and pretty sure there's tests that cover this.  Something weird about encoding range spans with null boundaries requires it.

I don't really understand what your change is doing, a logic test that shows the difference in the spans with some real values would be helpful.

**Review by msirek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @rytaft, and @yuzefovich)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N09782G8D0LPlDqvsE0:-N09782G8D0LPlDqvsE1:b-5k816n) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ba6bf4cc821550d6991cf900c4cd5528eae98aa/pkg/sql/rowexec/joinreader_span_generator.go#L831)):*
> ```Go
> func (g *localityOptimizedSpanGenerator) getMatchingRowIndices(key roachpb.Key) ([]int, error) {
> 	if res, err := g.localSpanGen.getMatchingRowIndices(key); len(res) > 0 || err != nil {
> 		return res, err
> ```

Could there be no spans containing the key in `localSpanGen`, but spans in `remoteSpanGen` which do contain the key?  If so, perhaps this shouldn't return here on an error, but fall through and do the lookup in `remoteSpanGen`.  It might be worthwhile to try out a test case.

**Review by yuzefovich — COMMENTED** [review]
Added tests showing the changes.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-:-N09lA0k6DrmL7UKFIeQ:b-7vaolj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, and @rytaft)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N09782G8D0LPlDqvsE0:-N09lRu5Dqa5rQ01uY6i:b-pcwy0a) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ba6bf4cc821550d6991cf900c4cd5528eae98aa/pkg/sql/rowexec/joinreader_span_generator.go#L831)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Could there be no spans containing the key in `localSpanGen`, but spans in `remoteSpanGen` which do contain the key?  If so, perhaps this shouldn't return here on an error, but fall through and do the lookup in `remoteSpanGen`.  It might be worthwhile to try out a test case.
</blockquote></details>

Nice catch! I think you're right, I'll try to come up with a test case.

**Comment by cucaroach** [comment]
Thanks for the test cases!  I think I thought my code already achieved this with the startKey.Next() call.  I guess for some types a key next is finer than a value next?  I vaguely remember playing with value.Next and seem to recall it didn't work for all datum types?  Maybe the reason startKey.Next() doesn't work is related to the reason a seemingly extra .Next call is needed for the nil case.  I think you might have to fallback to key next when value Next fails?   Although we'd want to test what happens on > MAX cases.  Relying on optimizer to turn that into false seems dicey to me but Rebecca would know better.

**Review by rytaft — APPROVED** [review]
:lgtm:

> Reviewed 1 of 3 files at r1, 4 of 4 files at r2, 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-:-N0EYz6r3b9ZGybvChp0:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach, @msirek, and @yuzefovich)

___
*[pkg/sql/span/span_builder.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N06d-rY6oFd4CMb6HAl:-N0EXl116YJ9Lmesd57D:blawk6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ba6bf4cc821550d6991cf900c4cd5528eae98aa/pkg/sql/span/span_builder.go#L108)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Curious if Becca or Tommy have any insight on this.
</blockquote></details>

I don't think the optimizer does this. Can you add a test case with this example to confirm? It can detect some contradictions, but not this kind as far as I know. We'll probably need to add a rule to detect cases like this.

(or feel free to open an issue)

**Review by yuzefovich — COMMENTED** [review]
> I think I thought my code already achieved this with the startKey.Next() call. I guess for some types a key next is finer than a value next?

Indeed, key next is working on the encoding whereas value next is working on the type values, and the latter is more precise.

> I vaguely remember playing with value.Next and seem to recall it didn't work for all datum types?

Yes, these methods don't work for Min/Max values of some types and for all values of other types (like decimals). I made the change to disable range-based lookup joins in all such cases, and I think this limitation is worth it - it'll allow us to get the improvements I mentioned [here](https://github.com/cockroachdb/cockroach/pull/80172#issuecomment-1105467943). It'll also unblock (or at least simplify) the work on the streamer.

I think this is RFAL.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-:-N0HzW6E1_vm78-ivSFe:be2wrz0)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @msirek, and @rytaft)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N09782G8D0LPlDqvsE0:-N0HzFtD4JbCzDFa1Af0:b-36b86z) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ba6bf4cc821550d6991cf900c4cd5528eae98aa/pkg/sql/rowexec/joinreader_span_generator.go#L831)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Nice catch! I think you're right, I'll try to come up with a test case.
</blockquote></details>

Indeed, this was a problem, and I added the tests for it.
___
*[pkg/sql/span/span_builder.go, line 108 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N06d-rY6oFd4CMb6HAl:-N0HtPbVEnwMcMUOgBAp:b9dvt0g) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ba6bf4cc821550d6991cf900c4cd5528eae98aa/pkg/sql/span/span_builder.go#L108)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I don't think the optimizer does this. Can you add a test case with this example to confirm? It can detect some contradictions, but not this kind as far as I know. We'll probably need to add a rule to detect cases like this.

(or feel free to open an issue)
</blockquote></details>

Filed #80402. I went around this by disabling range-based lookup joins when the datum cannot be "advanced".

**Review by msirek — APPROVED** [review]
:lgtm: 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-:-N0I1VFrEnRJ5gM1nZS5:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @rytaft, and @yuzefovich)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N09782G8D0LPlDqvsE0:-N0I0We72dwPdya1NpUb:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ba6bf4cc821550d6991cf900c4cd5528eae98aa/pkg/sql/rowexec/joinreader_span_generator.go#L831)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Indeed, this was a problem, and I added the tests for it.
</blockquote></details>

Thanks!

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @cucaroach, @msirek, and @rytaft)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N09782G8D0LPlDqvsE0:-N0I4V-j3Q329gbERtr5:bac1rl0) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ba6bf4cc821550d6991cf900c4cd5528eae98aa/pkg/sql/rowexec/joinreader_span_generator.go#L831)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

Thanks!
</blockquote></details>

Actually, I think we can improve here. Added another commit, PTAL.

**Review by msirek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @cucaroach, @msirek, @rytaft, and @yuzefovich)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 820 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N0IAtWy0i7MwhhLw6tn:-N0IAtWy0i7MwhhLw6to:b-i2129v) ([raw file](https://github.com/cockroachdb/cockroach/blob/31d3b6d4019b10282e1593c56c3193802da6f4be/pkg/sql/rowexec/joinreader_span_generator.go#L820)):*
> ```Go
> 	ctx context.Context, rows []rowenc.EncDatumRow,
> ) (roachpb.Spans, error) {
> 	g.localSpanGenUsedLast = true
> ```

I'm a little confused how this works.  For example, in this code:
```
	// Figure out what key spans we need to lookup.
	spans, err := jr.strategy.processLookupRows(jr.scratchInputRows)
	if err != nil {
		jr.MoveToDraining(err)
		return jrStateUnknown, nil, jr.DrainHelper()
	}
	jr.curBatchInputRowCount = int64(len(jr.scratchInputRows))
	jr.resetScratchWhenReadingInput = true
	jr.curBatchSizeBytes = 0
	jr.curBatchRowsRead = 0
	if len(spans) == 0 {
		// All of the input rows were filtered out. Skip the index lookup.
		return jrEmittingRows, outRow, nil
	}
```
`processLookupRows` calls `generateSpans`, which only processes the locals spans.  Then if there are no spans or an error, we don't try to use the remote spans.  What am I missing?

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @cucaroach, @msirek, and @rytaft)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 820 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N0IAtWy0i7MwhhLw6tn:-N0IIm29AbMUjBtz4P2C:b-2y84xc) ([raw file](https://github.com/cockroachdb/cockroach/blob/31d3b6d4019b10282e1593c56c3193802da6f4be/pkg/sql/rowexec/joinreader_span_generator.go#L820)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

I'm a little confused how this works.  For example, in this code:
```
	// Figure out what key spans we need to lookup.
	spans, err := jr.strategy.processLookupRows(jr.scratchInputRows)
	if err != nil {
		jr.MoveToDraining(err)
		return jrStateUnknown, nil, jr.DrainHelper()
	}
	jr.curBatchInputRowCount = int64(len(jr.scratchInputRows))
	jr.resetScratchWhenReadingInput = true
	jr.curBatchSizeBytes = 0
	jr.curBatchRowsRead = 0
	if len(spans) == 0 {
		// All of the input rows were filtered out. Skip the index lookup.
		return jrEmittingRows, outRow, nil
	}
```
`processLookupRows` calls `generateSpans`, which only processes the locals spans.  Then if there are no spans or an error, we don't try to use the remote spans.  What am I missing?
</blockquote></details>

The thing is that `generateSpans` always returns non-empty `spans` slice - that slice will contain all local spans. Then we perform the lookup. If the rows turn out to be placed in remote region, then all those local lookups will return nothing, and in `performLookup` we'll realize that not all input rows got a match, so we will generate all remote spans.

In other words, we always generate local spans and perform the lookup of them. Then if all input rows got a match locally, then we skip the remote span generation and lookup.

**Review by msirek — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @cucaroach, @msirek, and @rytaft)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 820 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N0IAtWy0i7MwhhLw6tn:-N0IMrmz60zMyYtdNjv9:blmme42) ([raw file](https://github.com/cockroachdb/cockroach/blob/31d3b6d4019b10282e1593c56c3193802da6f4be/pkg/sql/rowexec/joinreader_span_generator.go#L820)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

The thing is that `generateSpans` always returns non-empty `spans` slice - that slice will contain all local spans. Then we perform the lookup. If the rows turn out to be placed in remote region, then all those local lookups will return nothing, and in `performLookup` we'll realize that not all input rows got a match, so we will generate all remote spans.

In other words, we always generate local spans and perform the lookup of them. Then if all input rows got a match locally, then we skip the remote span generation and lookup.
</blockquote></details>

OK, but if `processLookupRows` returns no spans or an error, the above code seems to return immediately and proceed to state `jrEmittingRows` or `jrStateUnknown` directly and never enter the `jrPerformingLookup` state which triggers the call to `performLookup`.  If `performLookup` isn't called, then it looks like we'd miss calling `generateRemoteSpans` and perhaps miss some rows.



_Code quote:_
```Go
getMatchingRowIndices
```

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @cucaroach, @msirek, and @rytaft)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 820 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N0IAtWy0i7MwhhLw6tn:-N0IP-DE2gmfoBSp_kC7:b-ry2hwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/31d3b6d4019b10282e1593c56c3193802da6f4be/pkg/sql/rowexec/joinreader_span_generator.go#L820)):*
<details><summary><i>Previously, msirek (Mark Sirek) wrote…</i></summary><blockquote>

OK, but if `processLookupRows` returns no spans or an error, the above code seems to return immediately and proceed to state `jrEmittingRows` or `jrStateUnknown` directly and never enter the `jrPerformingLookup` state which triggers the call to `performLookup`.  If `performLookup` isn't called, then it looks like we'd miss calling `generateRemoteSpans` and perhaps miss some rows.
</blockquote></details>

`processLookupRows` will never return no spans (unless it returns an error, in which case the query execution should fail).

**Review by msirek — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @rytaft, and @yuzefovich)

___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 831 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N09782G8D0LPlDqvsE0:-N0IRKMN9Oz65LwvJ_WD:bg7igkv) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ba6bf4cc821550d6991cf900c4cd5528eae98aa/pkg/sql/rowexec/joinreader_span_generator.go#L831)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Actually, I think we can improve here. Added another commit, PTAL.
</blockquote></details>

:lgtm: 
___
*[pkg/sql/rowexec/joinreader_span_generator.go, line 820 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-N0IAtWy0i7MwhhLw6tn:-N0IR3wf6PVKpQzq4WG5:b-ptc25z) ([raw file](https://github.com/cockroachdb/cockroach/blob/31d3b6d4019b10282e1593c56c3193802da6f4be/pkg/sql/rowexec/joinreader_span_generator.go#L820)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

`processLookupRows` will never return no spans (unless it returns an error, in which case the query execution should fail).
</blockquote></details>

Oh, I see.  We've already buffered at least some rows at this point and are just building spans from them.  OK.

:lgtm:

**Review by yuzefovich — COMMENTED** [review]
Right. The execution flow is as follows:
- buffer a batch of input rows
- generate local spans for all of them
- perform lookup of local spans
- if not all rows got a match, generate remote spans for all input rows
- perform lookup of remote spans
- proceed to the first step of getting the next input rows batch.

TFTRs!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/80213#-:-N0IRg3u27i2HWq61lBF:b-ly9c2e)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach, @rytaft, and @yuzefovich)
